pkg/app/piped/analysisprovider/metrics/datadog/datadog.go

```diff
@@ -16,43 +16,135 @@ package datadog
 
 import (
 	"context"
+	"fmt"
+	"net/http"
 	"time"
 
+	"github.com/DataDog/datadog-api-client-go/api/v1/datadog"
+	"go.uber.org/zap"
+
 	"github.com/pipe-cd/pipe/pkg/config"
 )
 
-const ProviderType = "Datadog"
+const (
+	ProviderType   = "Datadog"
+	defaultAddress = "datadoghq.com"
+	defaultTimeout = 30 * time.Second
+)
 
-// Provider is a client for datadog.
+// Provider works as an HTTP client for datadog.
 type Provider struct {
-	metricsQueryEndpoint     string
-	apiKeyValidationEndpoint string
+	client *datadog.APIClient
 
-	timeout        time.Duration
-	apiKey         string
-	applicationKey string
-	fromDelta      int64
+	address           string
+	apiKey            string
+	applicationKey    string
+	queriedTimePeriod int64
+	timeout           time.Duration
+	logger            *zap.Logger
 }
 
-func NewProvider(address, apiKey, applicationKey string) (*Provider, error) {
-	return &Provider{
-		metricsQueryEndpoint: address,
-		apiKey:               apiKey,
-		applicationKey:       applicationKey,
-	}, nil
+func NewProvider(apiKey, applicationKey string, queriedTimePeriod time.Duration, opts ...Option) (*Provider, error) {
+	if apiKey == "" {
+		return nil, fmt.Errorf("api-key is required")
+	}
+	if applicationKey == "" {
+		return nil, fmt.Errorf("application-key is required")
+	}
+	if queriedTimePeriod == 0 {
+		return nil, fmt.Errorf("aggregation period is required")
+	}
+
+	p := &Provider{
+		client:            datadog.NewAPIClient(datadog.NewConfiguration()),
+		address:           defaultAddress,
+		apiKey:            apiKey,
+		applicationKey:    applicationKey,
+		queriedTimePeriod: int64(queriedTimePeriod.Seconds()),
+		timeout:           defaultTimeout,
```

Comment:
Instead of setting these `timeout` and `queriedTimePeriod` parameters as the shared values inside the provider scope,
maybe they should be passed when calling the query.
For example, in the case of `timeout`, it can be set via the parent content.

Reply:
Thank you for taking a look! There you go. Actually those parameters should be given whenever calling the query as you said.

But honestly to do so, we need to change this generic interface (Hopefully Go supports generic enum type right now but it's gonna be supported in a little bit more future), that's why I settled on putting those parameters as strust's fields.
https://github.com/pipe-cd/pipe/blob/0f129c7862c269bbe51b0411eed23f462c337cee/pkg/app/piped/analysisprovider/metrics/provider.go#L30-L35

But I feel the same way as you, so let me add a TODO comment (For `timeout`, we can address easily though)

Reply:
> Hopefully Go supports generic enum type

Not sure why do we need the Generic feature. That feature is needed for `queriedTimePeriod`?

Can we express the period by the query string? I mean including the period in the query itself. For example, `[1m]` for Prometheus. Users can configure to get a scalar value, vector, or matrix just by their query.

Reply:
>including the period in the query itself.

I was thinking about it as well to unify with Prometheus, but seems like we can't do that with Datadog. Looks like they are slightly different from each other, that's why I'm feeling it's a little bit difficult to unify the interface and seeking the Generic feature (Sorry to change the subject. The generics enum feature is not required, so never mind)

Reply:
>we check all data points not just the last point of the first series

There you go. You're quite right. Initially I was thinking users should devise a query so that the result would be a single value, but it's really tedious.

Also, `QueryParams` looks suitable for us. On second thought, `queriedTimePeriod` is not a specific kind of parameter at all, quite common for most providers. Let me look into QueryRange stuff, and thank you for your fancy feedback!

>about the query range, IMO it can be from the time the lastest query was sent to time.Now)

I feel this range is appropriate at any time.

Reply:
👍